Case study - Java functional programming using Vavr.

p2w34 opened the pull request.

p2w34 committed:
[draft] java with vavr

p2w34 committed:
[draft] Case study - Java FP using Vavr.

p2w34 committed:
[draft][grammar] Case study - Java FP using Vavr.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+# Case study: Java functional programming using Vavr
+
+Most Vavr tutorials show how to use Vavr primitives but they do not provide intuition on how to write whole applications.
```
Just a single sentence about what Vavr is + a link :)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+Just one remark here. One might argue whether the word 'pipeline' is appropriate or not.
+Why not simply 'stream'? I find 'pipeline' accurate as streams need to flow through something.
+From all such 'somethings' first thing which comes to my mind is 'pipeline'.
```
Oh ok, I see the picture below implies linear flow

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+They both present programs as pipelines. A single arrow should be interpreted as a single computational stage (for example a single method),
+whereas the colors mean:  
+- green - purely functional (no side effects),  
+- orange - with side effects (especially throwing errors),  
```
Hmm the mention about throwing errors here is a bit confusing (esp. wrt. the distinction b/w red and orange)... but I hope it'll get clearer down the way

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+The left example shows a purely functional program, meaning none of the pipeline stages performs any side effects.
+It is hard to imagine that such a program could be of any use - no side effects means no printing result on the screen/
+writing to a file, no possibility to input parameters, etc.  
```
Wrt. `input parameters`: that's a bit unclear coz a program can get parameters at the beginning and still remain purely functional... pls specify what you mean here (e.g. reading extra input from the file or providing parameters interactively during the execution)?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+The example on the right is more interesting - apart from doing purely FP-style calculations,
+it reads user input, calls database service and can display some results on the screen.
+Should any error occur, it will be wrapped (with help of Vavr) into a proper type and returned (not thrown) from a method.
```
`s/with help/with the help`

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+It depicts our microservice as a pipeline.
+Numbers marking aspects to be examined are added as well:
+
+![service_flow](images/service_flow.png)
```
Hmm I'd consistently use verbs and not nouns (like `DB Service` or `Security token`) to name the orange boxes (pipeline stages).

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+it is useful for both the next service run as well to know from where to continue after crashes.
+
+The exact algorithm describing the service functionality is as follows:
+- get last 'polling status' from S3
```
`s/last/latest` (more unambiguous)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+Let's take a look at the class responsible for publishing events to the external endpoint through REST.
+It shows how to wrap an HTTP call which might throw an exception.
+Should an exception be thrown, it will be wrapped into Try, then translated to Either.Left and returned from the method.
+(note: of course, when you base your pipeline on Try, conversion to Either is not necessary).
```
Maybe you could factor out `Try` completely to simplify the narrative?

PawelLipski replied:
Or only keep it inside method implementation where needed, but instantly convert to Either? i.e. never exposing it from any method

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+Let's assume external HTTP calls are done by client implementing the following interface:
+```
+interface InventoryPublisherClient {
+    Boolean publishItem(ItemData item) throws RuntimeException;
````
Hmm does it make sense to declare `throws` for `RuntimeException`? In a sense, every method in Java `throws RuntimeException` by default ;)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+Let's talk about some basics first.  
+Vavr equips us with tools one may find valuable writing such loops.
+These are primitives like Stream.continually(), takeWhile(), dropWhile() and find().
```
Nit: format method names as `code`

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+And their 'brothers and sister' (takeUntil() etc.).  
+
+**Now comes a very important thing, crucial to understanding this paragraph.
+The io.vavr.collection.Stream implementation is a lazy linked list.**
```
Btw this list is unidirectional, right (you can only go forward in the stream but never backward)? Maybe worth mentioning if so

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+@Test
+void testFind() {
+    var resultStream = testStream.map(
+            pr -> anyServiceStream.serviceCall(pr) # it returns false for the first three calls, true for the rest
```
`#` for comments? ;)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+        .find(processingResult -> processingResult.isResult() == false);
+
+    var resultList = testList.map(
+            pr -> anyServiceList.serviceCall(pr) # it returns false for the first three calls, true for the rest
```
`anyServiceList::serviceCall`?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+    var resultList = testList.map(
+            pr -> anyServiceList.serviceCall(pr) # it returns false for the first three calls, true for the rest
+        )
+        .find(processingResult -> processingResult.isResult() == false);
```
`.find(processingResult -> !processingResult.isResult())`? Or even `.findNot(processingResult::isResult)` if such method exists

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+    var resultStream = testStream.map(
+            pr -> anyServiceStream.serviceCall(pr) # it returns false for the first three calls, true for the rest
+        )
+        .find(processingResult -> processingResult.isResult() == false);
```
`isResult`... hmm what does this method return actually? ;D what does it mean when `processingResult IS NOT result`? ;)

PawelLipski replied:
Maybe `isSuccessful` or sth like that instead?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+}
+```
+
+One might observe that lazy implementation of Stream class causes that anyServiceStream.serviceCall(pr)
````
`code` formatting

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+                return Either.left(fetchResult.getLeft());
+            }
+
+            availableItemsResponse = fetchResult.get();
```
Make it clear somehow that `get()` returns the Right part (maybe `.getRight()` if exists?)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+                return Either.left(publishResult.getLeft());
+            }
+
+        } while (availableItemsResponse.getHasMore() == true);
```
Remove `== true`

PawelLipski replied:
Plus rename the method to `hasMore` if under your control

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+        return Stream.continually(() -> fetchAndPublishAvailableItems())
+                .find(this::isLastPage)
+                .get()
+                .map(__ -> true);
```
Maybe put a comment that `__` name indicates that the param is to be ignored

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+```
+public static <T> Either<Throwable, T> logInfo(T rightValue, String message) {...}
+```
+and then plugged into the pipeline using map() instead of peek().
````
Explain the difference b/w the effect of `peek` and `map`

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+```
+public static void main(String[] args) {
+    Try.of(() -> startApplication())
+        .onFailure(t -> logError(t, "Service failuer"))
````
Typo

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+one might refactor it to:
+```
+public boolean filter2(String label) {
+    return Option.of(label)
````
`ofNullable`?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+    return Option.of(label)
+            .map(String::toUpperCase)
+            .filter(allowedLabels::contains)
+            .map(__ -> true)
```
`.map(__ -> true).getOrElse(false)` => `.isPresent()`?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+# Summary
+The case study presented above should enable the reader to write Java microservices in a functional style using Vavr.
+It may also serve as proof that it is not necessary to even mention (in)famous 'm-word' to talk about FP usage in practice!
```
Hmmm I don't think the joke is going to be very clear... esp. to the people who, well, don't operate on monads on a daily basis (and I think that might be the target group) ;) what adds to the confusion is proximity of another m-word `microservices` :/

ArturSkowronski replied:
I agree with you there, but I have one question to @p2w34 - who is really a target of that post? Cause after reading it, I see that it will appeal to people who like FP, need to write in Java for some reason and want to use Vavr to emulate some techniques. Did I recognize correctly? Cause for newbies, I lack a bit of a better explanation of how this approach helps.

p2w34 replied:
Joke removed.

@ArturSkowronski - I am very happy seeing a comment like this. It is clear advice that I should modify Intro/Summary.
This post is meant for regular (regular meaning that on a daily basis they write Java code, it is not about seniority level) Java developers, who heard/even bought books about FP in Java, they might have used Vavr here and there, but still they do not know how to combine this knowledge so that they write entire applications in this style. And I am writing it based on my own experience - I have never used any Scala/Haskell, etc. But I do see the benefits of FP and I would like to share what I already know, in the simplest possible form.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/code-examples/src/test/java/com/pbroda/codesnippets/vavr/LoopCaseStudy.java:
```diff
@@ -0,0 +1,217 @@
+    Either<Throwable, Boolean> publishAvailableItems(List<Item> items);
+}
+
+//@Value
```
Remove or uncomment

PawelLipski reviewed (changes requested):
Good job! review done

ArturSkowronski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,338 @@
+However, one might be tempted to write logging methods such that (internally) they call other methods with side effects.
+Example signature of such method would be:
+```
+public static <T> Either<Throwable, T> logInfo(T rightValue, String message) {...}
````
Isn't that a bit contrived example? It's hard for me to imagine somebody writing logic like that.

p2w34 replied:
It is a real example ;/. That is why I wanted to mention here that it is better to avoid such code. Or in your opinion, I should rather remove it?

ArturSkowronski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+after each such successful operation, update 'polling status' on S3.
+
+This example is based on a real microservice.
+It is built without any frameworks apart from Guice to provide dependency injection, but this is irrelevant here.
```
I'd get rid of that sentence. It's nice that you built microservice without frameworks, but as you said - it's irrelevant there

ArturSkowronski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+This is acceptable - thanks to storing 'polling status' on S3, the next run will continue from the place where previously broke).
+
+All external services are exposed through REST, but it does not matter here.
+What is important is that contacting them is a side effect.
```
Maybe merge those two sentences?

'All external services are exposed through REST, but what truly matter is that contacting them is a side effect.'

ArturSkowronski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+
+## Initial considerations
+Vavr equips us with two ways of building pipelines: classes Either and Try.
+Since Try might be treated as a specific case of Either, for the needs of this case study it does not really matter which one we choose.
```
Maybe stick to one then?

ArturSkowronski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    private Boolean isLastPage(Either<Throwable, Tuple2<Boolean, Boolean>> fetchAndPublishResult) {
+        return fetchAndPublishResult.isLeft() ||
```
This code is basically ugly to somebody who isn't already fan of FP approach. It's a bit preaching to the already converted. Maybe that the goal.

p2w34 replied:
I am not sure whether I understand your comment. For most people, the code before refactoring is for sure more clear, but I wanted to present how one can write it using Vavr the proper way.

p2w34 committed:
[draft][apply review comments] Case study - Java FP using Vavr.

p2w34 reviewed (commented):
Thank you for your comments. I hope that I was able to apply your suggestions/reply to the comments.
I will have to yet address Artur's question - let me think about it more.

p2w34 committed:
[draft][apply review comments - fixed images] Case study - Java FP using Vavr.

p2w34 committed:
[draft] change introduction + minor text corrections

p2w34 commented:
@ArturSkowronski - I just changed the introduction - I added information about the target audience.
Do you think that now it clarifies the whole idea behind this post?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+Let's take a look at the class responsible for publishing events to the external endpoint through REST.
+It shows how to wrap an HTTP call which might throw an exception.
+Should an exception be thrown, it will be wrapped into ```Try```, then translated to ```Either.Left``` and returned from the method.
+(note: of course, when you base your pipeline on ```Try```, conversion to ```Either``` is not necessary).
````
Side note: IMO `Try` generally sucks (same in Scala) coz (unlike `Either`) it doesn't allow for specifying upper bound on the exception type... just `Throwable` is all we've got :hankey:

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+Let's assume external HTTP calls are done by client implementing the following interface:
+```
+interface InventoryPublisherClient {
+    Boolean publishItem(ItemData item);
````
Nit: primitive `boolean` instead?

p2w34 replied:
Could be boolean as well, but I think it may stay as it is.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+```
+Then our wrapper class may simply take the form of:
+```
+@Value
````
Hmm where's that `@Value` coming from? Maybe it can be safely skipped?

PawelLipski replied:
Note that publishing code snippets (as well as tables) in Medium articles sucks a big time (just as many things about Medium's usability): I ended up cooking up Gists for mine recently: https://medium.com/virtuslab/nifty-docker-tricks-for-your-ci-vol-1-c4a36d2192ea

p2w34 replied:
Good to know - sounds like a reasonable way of tackling it.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+Let's talk about some basics first.  
+Vavr equips us with tools one may find valuable writing such loops.
+These are primitives like ```Stream.continually()```, ```takeWhile()```, ```dropWhile()``` and ```find()```.
+And their 'brothers and sister' (```takeUntil()``` etc.).  
````
```suggestion
And their 'brothers and sisters' (```takeUntil()``` etc.).  
```

p2w34 replied:
done

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+[here](./code-examples/src/test/java/com/pbroda/codesnippets/vavr/LoopBasics.java).
+
+Now it is time for our example. 
+Assume, we want to call two services in a loop (this correspond to the bigger rectangle on the picture).
```
```suggestion
Assume we want to call two services in a loop (this corresponds to the bigger rectangle on the picture).
```

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+Now it is time for our example. 
+Assume, we want to call two services in a loop (this correspond to the bigger rectangle on the picture).
+We call them consecutively in such a way that the result of the first call is fed to the second call.
+This proceeds in a loop until there is a certain condition met, based on the results of the first call.
```
```suggestion
This proceeds in a loop until there a certain condition is met, based on the results of the first call.
```

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+Let's repeat as it is important - first call, not the second one.
+
+If it is not obvious how to write such a loop, we might first try to write it as a kind of pseudo-code presented below.
+Unit tests not only assure us the functionality is correct, but also allow us with experimenting on the way to get a nice result.
```
```suggestion
Unit tests not only assure us the functionality is correct, but also allow to experiment on the way to get the desired result.
```

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+Unit tests not only assure us the functionality is correct, but also allow us with experimenting on the way to get a nice result.
+So our first version might look like: 
+```
+@Value
````
Again, `@Value` is unclear here

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+In order to refactor ```ProcessingPipeline``` class above we need to deal with two problems, both related to the condition
+which terminates the loop:  
+- it is based on the result of a service call - we have already covered it at the beginning of this paragraph,  
+- it is based on the result of the first service call, as already stressed.
````
Hmmm this second point is a bit a awkward - maybe sth along the lines of `another service call follows in each iteration after the service call that ...`

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+```   
+
+In order to refactor ```ProcessingPipeline``` class above we need to deal with two problems, both related to the condition
+which terminates the loop:  
````
`loop termination condition`?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+        return Stream.continually(() -> fetchAndPublishAvailableItems())
+                .find(this::isLastPage)
+                .get()
+                .map(ignore -> true);
```
Nit: `__ -> true` maybe, seems that's kinda convention

p2w34 replied:
As I already wrote somewhere, ignore is also fine so I leave it as it is.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+
+    private Either<Throwable, Tuple2<Boolean, Boolean>> fetchAndPublishAvailableItems() {
+        return warehouseService.fetchAvailableItems()
+                .flatMap(itemsResponse -> publishAvailableItems(itemsResponse));
```
```suggestion
                .flatMap(this::publishAvailableItems);
```

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+[here](./code-examples/src/test/java/com/pbroda/codesnippets/vavr/LoopCaseStudy.java).
+
+## Aspect 3 - logging
+When logging methods we write are meant to just log messages passed to them, i.e. their signatures are of form:
```
`When logging methods we write are meant to just log` - this part is unclear

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+in a consistent way.
+And that becomes hard for writing ```logError()``` method.
+```LogError()``` method is supposed to log exceptions, but it would be plausible that ```logError()``` itself would result in an exception.
+It would be still feasible to code it, with for example help of 'suppressed' exceptions, but one might unnecessarily
````
What do you mean specifically by `suppressed` exceptions? Wrapped into `RuntimeException`?

PawelLipski replied:
Wow ok, I didn't know this stuff previously :)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+In our example, we are in a very comfortable situation in which we allow them to propagate to the very end and log them there - just in one place.
+Should any translation of an error be needed at any stage of the pipeline, then simply:
+```
+.mapLeft(err -> translateError(err))
````
`::`?

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+Exceptions simply travel through the pipeline as ```Either.Left```.
+As they make their way towards the end of the pipeline, there might be a need for translating them.
+In our example, we are in a very comfortable situation in which we allow them to propagate to the very end and log them there - just in one place.
+Should any translation of an error be needed at any stage of the pipeline, then simply:
````
It isn't clear what you specifically mean by `translation`, pls specify

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+It would be still feasible to code it, with for example help of 'suppressed' exceptions, but one might unnecessarily
+end up here with an overcomplicated code.
+
+## Aspect 4 - exceptions
```
It might be a good idea to split the post into two parts, it's pretty long

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,335 @@
+.mapLeft(err -> translateError(err))
+```
+
+What is more, I also consider a good practice wrapping the starting point of an application with ```Try```:
````
```suggestion
What is more, I also consider it a good practice to wrap the starting point of the application into ```Try```:
```

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+public boolean filter2(String label) {
+    return Option.of(label)
+            .map(String::toUpperCase)
+            .filter(allowedLabels::contains)
```
`.exists(allowedLabels::contains)`? or sth like that

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+public Either<Throwable, Boolean> filter3(String label) {
+    return Option.of(label)
+            .map(allowedLabelsService::isAllowed)
+            .getOrElse(() -> Either.right(false));
```
`io.vavr.Option#toEither(L)` instead?

p2w34 replied:
`isAllowed() `already returns `Either`.

`.getOrElse()`, and '`Else`' part in particular is for scenario in which `Option.of(label)` returns `None`.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/code-examples/src/test/java/com/pbroda/codesnippets/vavr/ExternalServiceCall.java:
```diff
@@ -0,0 +1,80 @@
+}
+
+@ExtendWith(MockitoExtension.class)
+public class ExternalServiceCall {
```
`ExternalServiceCallTest` for clarity that it's test code?

p2w34 replied:
Done

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/code-examples/src/test/java/com/pbroda/codesnippets/vavr/LoopBasics.java:
```diff
@@ -0,0 +1,128 @@
+}
+
+@ExtendWith(MockitoExtension.class)
+public class LoopBasics {
```
Again, `LoopBasicsTest`?

mpociecha commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+*TL;DR;* This post is aimed mainly at Java programmers who haven't really had a chance to write their applications in a functional style, not even in other languages.
+It is based on [Vavr](https://vavr.io) which is a popular functional library for Java 8+ and assumes basic familiarity with it.
+Presented analysis should be seen as additional reading material to both Vavr's documentation and books/tutorials on Java functional programming (FP).
+However, to not scare the reader off - a basic understanding of FP concepts like immutability or pure functions is sufficient.
```
Given you started with comma, it should be `reader off, a basic`

mpociecha commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+
+They both present programs as pipelines. A single arrow should be interpreted as a single computational stage (for example a single method),
+whereas the colors mean:  
+- green - purely functional (no side effects),  
```
Nitpick: In such cases like this `green - purely` there should be used dashes instead of hyphens (see https://www.grammarly.com/blog/hyphens-and-dashes/)

mpociecha commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+
+![example_flows](images/example_flows.png)
+
+They both present programs as pipelines. A single arrow should be interpreted as a single computational stage (for example a single method),
```
It could be just `An arrow should be interpreted as a computational stage (for example a single method)`

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,335 @@
+(OK, some failed HTTP calls might be retried two or three times, but if this fails the pipeline breaks.
+This is acceptable - thanks to storing 'polling status' on S3, the next run will continue from the place where previously broke).
+
+All external services are exposed through REST, but what truly matters is that contacting them is a side effect.
```
I admit it is a more sophisticated word but I leave it as it is.

p2w34 committed:
[draft] Apply review comments

p2w34 committed:
[draft] remove redundant word (this)

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+- knowing that database response might consist of multiple pages, for each page (large loop on the picture):
+    - request access token and (small loop):
+        - publish to external endpoint (using access token) all entries from a single page, one at a time;
+after each such successful operation, update 'polling status' on S3.
```
`- ` at the start? The intended indent (no pun intended xD) is current unclear here

p2w34 replied:
It should be as it is. I hope it will look good on Medium.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+This schema is universal to all throwing methods.
+Once again - just wrap them before plugging into the pipeline.  
+Code with unit tests is available
+[here](./code-examples/src/test/java/com/pbroda/codesnippets/vavr/ExternalServiceCall.java).
```
Update the links (`...Test.java`)

p2w34 replied:
Good catch!

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,333 @@
+
+Let's talk about some basics first.  
+Vavr equips us with tools one may find valuable writing such loops.
+These are primitives like ```Stream.continually()```, ```takeWhile()```, ```dropWhile()``` and ```find()```.
+And their 'brothers and sisters' (```takeUntil()``` etc.).  
````
```suggestion
These include primitives like ```Stream.continually()```, ```takeWhile()```, ```dropWhile()```, ```find()``` and their counterparts like ```takeUntil()``` etc.
```

p2w34 replied:
:1st_place_medal:

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,333 @@
+will be called as many times as needed to fulfill ```find()``` condition. 
+This is not the case for the ```List``` - ```anyServiceList::serviceCall``` will be executed as many times as there are elements in ```testList```.
+Unit tests showing not only ```find()``` behavior but also other mentioned methods can be found
+[here](./code-examples/src/test/java/com/pbroda/codesnippets/vavr/LoopBasics.java).
````
`...Test.java`

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,333 @@
+Apart from sticking to single responsibility rule, it would be great to have methods to log all levels (info, debug, error, etc.)
+in a consistent way.
+And that becomes hard for writing ```logError()``` method.
+```LogError()``` method is supposed to log exceptions, but it would be plausible that ```logError()``` itself would result in an exception.
````
```suggestion
```logError()``` method is supposed to log exceptions, but it would be plausible that ```logError()``` itself would result in an exception.
```

p2w34 replied:
Done

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,333 @@
+## Aspect 4 - exceptions
+Exceptions simply travel through the pipeline as ```Either.Left```.
+As they make their way towards the end of the pipeline, there might be a need for translating them.
+In our example, we are in a very comfortable situation in which we allow them to propagate to the very end and log them there - just in one place.
````
(?)
```suggestion
In our example, we are in a very comfortable situation in which we allow them to propagate to the very top and log them there - just in one place.
```

p2w34 replied:
OK, top might be better - applied.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
````diff
@@ -0,0 +1,333 @@
+Exceptions simply travel through the pipeline as ```Either.Left```.
+As they make their way towards the end of the pipeline, there might be a need for translating them.
+In our example, we are in a very comfortable situation in which we allow them to propagate to the very end and log them there - just in one place.
+Should any translation (rethrowing) of an error be needed at any stage of the pipeline, then simply:
````
```suggestion
Should any translation (wrapping & rethrowing) of an error be needed at any stage of the pipeline, then simply:
```

p2w34 replied:
I admit I haven't thought it too well when I added 'rethrowing'. The point is that in Vavr instead of throwing exceptions we use algebraic types. I change it to:
`Should any translation (by analogy to pure Java _rethrowing_ ) of an error be needed at any stage of the pipeline, then simply.`
Also, wrapping is not mandatory in such cases.

PawelLipski commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+
+Which is better? IMHO it depends:
+- if there are no side effects involved, just pure logic like in the example above, then it is a matter of personal taste,  
+- if there are side effects involved, I would compose code into a pipeline as this allow to propagate exceptions nicely:  
```
```suggestion
- if there are side effects involved, I would compose code into a pipeline as this allows to propagate exceptions nicely:  
```

PawelLipski replied:
Run the blogpost through Grammarly or sth like that btw

p2w34 replied:
Did it but at this stage only very few minor corrections are needed.

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+In FP errors are not thrown as they would be in pure Java. Instead, they are returned from methods similar to non-error values.
+This allows us to visualize programs as pipelines, through which errors flow alongside non-error results.  
+
+Just one remark here. One might argue whether the word 'pipeline' is appropriate or not.
```
I applied the general remark. This paragraph might seem to be unnecessary but as the idea of the pipeline is at the core of this post, I'd like to leave it as it is.

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+
+The example on the left shows a purely functional program, meaning none of the pipeline stages performs any side effects.
+It is hard to imagine that such a program could be of any use - no side effects means no printing result on the screen/
+writing to a file, no possibility to interactively input parameters during execution, etc.  
```
Grammarly recommends comma before etc. - leaving as it is.

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+The example on the right is more interesting - apart from doing purely FP style calculations,
+it reads user input, calls database service and can display some results on the screen.
+Should any error occur, it will be wrapped (with the help of Vavr) into a proper type and returned (not thrown) from a method.
+The pipeline, assembled from Vavr primitives, takes care of proper error propagation to the place where the program begins.
```
+1, but with `entrypoint` written as `entry point`.

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+Should any error occur, it will be wrapped (with the help of Vavr) into a proper type and returned (not thrown) from a method.
+The pipeline, assembled from Vavr primitives, takes care of proper error propagation to the place where the program begins.
+
+Writing programs in FP fashion is nothing more complicated than that.
+It is about assembling such pipelines and paying attention to side effects.
```
Applied

p2w34 commented on 2020-03-21-case-study-java-fp-using-vavr/case-study-java-fp-using-vavr.md:
```diff
@@ -0,0 +1,333 @@
+Writing programs in FP fashion is nothing more complicated than that.
+It is about assembling such pipelines and paying attention to side effects.
+Now we are ready to analyze a slightly more complicated program.
+
```
done

p2w34 committed:
[draft] apply review comments

p2w34 committed:
[draft] fix indentation
